### PR TITLE
Connectd does demux part 1

### PR DESCRIPTION
**Pull request by rustyrussell**
~(Builds on #4972)~ Rebased on master

This puts connectd in the middle of every peer connection: it does the decryption/encryption and streams the gossip_store.

Changelog-None

### CODE REVIEW

**Review by cdecker — COMMENTED** [review]
Preliminary comments I had collected before the end of year.

File: connectd/multiplex.c [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+			 read_from_peer(peer_conn, peer),
+			 write_to_peer(peer_conn, peer));
+}
+
+void multiplex_final_msg(struct peer *peer, const u8 *final_msg TAKES)
+{
+	peer->final_msg = tal_dup_talarr(peer, u8, final_msg);
+	if (!peer->to_subd)
+		io_wake(peer->peer_outq);
+}
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply connectd/multiplex.c:546]: We had a traditional "send and error and disconnect" in openingd, this just moves it. I'll have to see if this is warranted in the end...

File: common/gossip_store.c [hunk 2]
```diff
@@ -77,73 +63,52 @@ u8 *gossip_store_iter(const tal_t *ctx,
...
 
 		if (checksum != crc32c(be32_to_cpu(hdr.timestamp), msg, msglen))
 			status_failed(STATUS_FAIL_INTERNAL_ERROR,
-				      "gossip_store: bad checksum offset %"
-				      PRIi64": %s",
-				      off ? (s64)*off :
-				      (s64)lseek(*gossip_store_fd,
-						 0, SEEK_CUR) - msglen,
-				      tal_hex(tmpctx, msg));
+				      "gossip_store: bad checksum offset %zu"
```
Comment (cdecker) [inline common/gossip_store.c:89]: `bad checksum offset` -> `bad checksum at offset`

File: connectd/multiplex.c [hunk 3]
```diff
@@ -49,102 +71,102 @@ static struct io_plan *write_to_peer(struct io_conn *peer_conn,
...
+static struct io_plan *read_body_from_peer_done(struct io_conn *peer_conn,
+						struct peer *peer)
+{
+       u8 *decrypted;
+
+       decrypted = cryptomsg_decrypt_body(NULL, &peer->pps->cs,
+					  peer->peer_in);
+       if (!decrypted)
+               return io_close(peer_conn);
+       tal_free(peer->peer_in);
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply connectd/multiplex.c]: Yes, in fact there was a fixme in the old code related to this lack of logging.

File: connectd/multiplex.h [hunk 4]
```diff
@@ -38,4 +38,8 @@ struct io_plan *multiplex_peer_setup(struct io_conn *peer_conn,
 /* Send this message to peer and disconnect. */
 void multiplex_final_msg(struct peer *peer,
 			 const u8 *final_msg TAKES);
+
+/* Inject a message into the output stream */
+void queue_peer_msg(struct peer *peer, const u8 *msg TAKES);
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply connectd/multiplex.h:25]: Yes, this name was taken from gossipd.c, so less code had to be rewritten, but now it's a bit weird-looking.  Will rename later: eventually multiplex.[ch] should be renamed to peer.[ch].

File: connectd/multiplex.c [hunk 5]
```diff
@@ -73,6 +122,14 @@ static struct io_plan *encrypt_and_send(struct peer *peer,
 	}
 #endif
 
+	/* These are critical, and so send without delay. */
+	if (type == WIRE_COMMITMENT_SIGNED
+	    || type == WIRE_REVOKE_AND_ACK
+	    || type == WIRE_PING
+	    || type == WIRE_PONG) {
+		set_urgent_flag(peer, true);
+	}
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply connectd/multiplex.c:252]: It's actually cached in the peer flag, so optimal.  But will switch to um, switch.

File: wire/wire_io.h [hunk 6]
```diff
@@ -27,7 +27,7 @@ struct io_plan *io_read_wire_(struct io_conn *conn,
 
 /* Write message from data (tal_count(data) gives length).  data can be take() */
 struct io_plan *io_write_wire_(struct io_conn *conn,
-			       const u8 *data,
+			       const u8 *data TAKES,
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply wire/wire_io.h:30]: Yes, indeed.

File: connectd/multiplex.c [hunk 7]
```diff
@@ -101,6 +158,10 @@ static struct io_plan *write_to_peer(struct io_conn *peer_conn,
 						peer->final_msg,
 						after_final_msg);
 		}
+
+		/* Reset urgent flag, if set. */
+		set_urgent_flag(peer, false);
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply connectd/multiplex.c]: Actually, yes, this is wrong.  It should *only* be called in encrypt_and_send.

File: connectd/connectd.c [hunk 8]
```diff
@@ -446,11 +444,9 @@ struct io_plan *peer_connected(struct io_conn *conn,
 	if (!peer)
 		return io_close(conn);
 
-	/* FIXME: Remove pps abstraction! */
-	pps = new_per_peer_state(tmpctx);
-
 	/* If gossipd can't give us a file descriptor, we give up connecting. */
-	if (!get_gossipfds(daemon, id, their_features, pps)) {
+	gossip_fd = get_gossipfd(daemon, id, their_features);
+	if (gossip_fd < 0) {
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply connectd/connectd.c:447]: gossipd already logs broken here (and it knows the errno): adding comment here.

[the remark this answers is not in the document]
Comment (rustyrussell) [reply connectd/connectd.c:447]: Wait, get_gossipd also logs broken...

**Comment by rustyrussell** [comment]
OK, last three commits are new, as is https://github.com/ElementsProject/lightning/pull/4984/commits/264ae921f1af62dce6cd2f751ee0efdee1d10ec2 which is a fixup.

**Comment by cdecker** [comment]
Excellent changes, just needs squashing :-)

<details>
```
git range-diff $(git merge-base ef04f77735a159e042e5e05cbf4f836388f02a40 origin/master)..ef04f77735a159e042e5e05cbf4f836388f02a40 $(git merge-base 502fe1064228317c87979533e77338f102217a12 origin/master)..502fe1064228317c87979533e77338f102217a12
 1:  663b3f983 =  1:  663b3f983 update mocks
 2:  cb19d2924 =  2:  cb19d2924 connectd: keep timeout timer around so we can disable it.
 3:  66bc45a68 =  3:  66bc45a68 connectd: maintain connection with peer, shuffle data.
 4:  c19d2596e =  4:  c19d2596e gossipwith: create our own internal sync_crypto functions.
 5:  90d467ab3 !  5:  44a498920 lightningd: make sure gossipd knows before we update blockheight.
    @@ gossipd/gossipd.c: static void new_blockheight(struct daemon *daemon, const u8 *
      }

      #if DEVELOPER
    -@@ gossipd/gossipd.c: static void handle_txout_reply(struct daemon *daemon, const u8 *msg)
    -   good = handle_pending_cannouncement(daemon, daemon->rstate,
    -                                       &scid, sat, outscript);
    -
    -+
    -   /* If we looking specifically for this, we no longer are. */
    -   remove_unknown_scid(daemon->seeker, &scid, good);
    -
     @@ gossipd/gossipd.c: static struct io_plan *recv_req(struct io_conn *conn,
        case WIRE_GOSSIPD_DEV_COMPACT_STORE_REPLY:
        case WIRE_GOSSIPD_GOT_ONIONMSG_TO_US:
 6:  947e054d0 =  6:  904426100 connectd: do decryption for peers.
 7:  56a53d736 =  7:  c73e2efd0 peer_io: replace crypto_sync in daemons, use normal wire messages.
 8:  8bdba3c96 =  8:  04927f198 per_peer_state: remove struct crypto_state
 9:  dedcef40e =  9:  70d52263d connectd: do dev_disconnect logic.
10:  18942d63c = 10:  c570252f6 connectd: do nagle by packet type.
 -:  --------- > 11:  264ae921f fixup! connectd: do nagle by packet type.
11:  217b63aae = 12:  479ed13a7 common: add routine for absolute timeouts (vs. relative).
12:  7ff1a3653 ! 13:  f354c9a22 connectd: serve gossip_store file for the peer.
    @@ connectd/multiplex.c: static struct io_plan *write_to_peer(struct io_conn *peer_
                                                after_final_msg);
                }

    --          /* Reset urgent flag, if set. */
    --          set_urgent_flag(peer, false);
    +-          /* Tell them to read again, */
    +-          io_wake(&peer->subd_in);
     +          /* If they want us to send gossip, do so now. */
     +          msg = maybe_from_gossip_store(NULL, peer);
     +          if (!msg) {
    -+                  /* Reset urgent flag, if set. */
    -+                  set_urgent_flag(peer, false);
    -
    --          /* Tell them to read again, */
    --          io_wake(&peer->subd_in);
     +                  /* Tell them to read again, */
     +                  io_wake(&peer->subd_in);

13:  a93ea224f = 14:  95727c0f3 subdaemons: don't stream gossip_store at all.
14:  73b996976 = 15:  644245d27 lightningd: remove per_peer_state struct.
15:  4eac6867e = 16:  6e6ae9395 connectd: remove per_peer_state in favor of keeping gossip_fd directly.
16:  e7ddcc32d = 17:  f2d66807b connectd: put more stuff into struct gossip_state.
17:  ef04f7773 = 18:  2a1317d16 connectd: get addresses from lightningd, not gossipd.
 -:  --------- > 19:  dc94ea569 various: minor cleanups from Christian's review.
 -:  --------- > 20:  a8ef64b04 connectd: drop support (unused) for @ during handshake.
 -:  --------- > 21:  502fe1064 connectd: implement @ correctly.
```
</details>

ACK https://github.com/ElementsProject/lightning/pull/4984/commits/502fe1064228317c87979533e77338f102217a12

**Comment by rustyrussell** [comment]
Fixup folded.

Ack https://github.com/ElementsProject/lightning/commit/b0b70533d1d99d13549e55cb6c96e1122fcde3cc

**Comment by cdecker** [comment]
Shepherding CI, otherwise GTG

ACK https://github.com/ElementsProject/lightning/pull/4984/commits/4bb4ace07d9b7abf581e934b585c39d5e2ffe0cb

**Comment by vincenzopalazzo** [comment]
I think there is only one missing that the Github view hide, and it is at line `4056` of `channeld.c`, we miss an API upgrade `msg_queue_new(peer, true);`

We miss at list another CI run

**Comment by cdecker** [comment]
Changes look good, just a couple of flaky tests:

 - `test_closing_specified_destination` gets stuck on `l1.rpc.call('close', {'id': chan13, 'destination': addr})`. `close` doesn't seem to want to return a result.
 - `test_htlc_rexmit_while_closing` and `test_penalty_outhtlc` seem to be killing `openingd` instead of reporting an `Unacceptable upfront_shutdown_script`. The outcome is the same (no channel) but the error message was a bit more informative.

**Comment by rustyrussell** [comment]
OK, I gave in and included many of the cleanups I had hoped to defer to a followup PR.  Let's see if they help or hinder the test failures.

**Comment by rustyrussell** [comment]
That... did not help :(

Since I cannot reproduce this locally, I'll be throwing debugging patches around...

**Comment by vincenzopalazzo** [comment]
>Since I cannot reproduce this locally, I'll be throwing debugging patches around...

Maybe this is a good example to introduce the docker image in the CI, in this way we can have the opportunity to run the same tests env locally, an example can be the following PR #4987. Maybe can be somethings that we can discuss

**Comment by rustyrussell** [comment]
NO SLEEP TIL CI PASSES...

**Comment by rustyrussell** [comment]
WHACK MOLE. WHACK...

**Review by vincenzopalazzo — APPROVED** [review]
ack https://github.com/ElementsProject/lightning/pull/4984/commits/886bedea7be511cd158ec5623a57d1203f8144c3 for me because I think that the following error is unrelated. There is only some slow run here

```
=================================== FAILURES ===================================
_________________________ test_upgrade_statickey_fail __________________________
[gw5] linux -- Python 3.6.15 /opt/hostedtoolcache/Python/3.6.15/x64/bin/python3

node_factory = <pyln.testing.utils.NodeFactory object at 0x7f5be9cfd1d0>
executor = <concurrent.futures.thread.ThreadPoolExecutor object at 0x7f5be9cfd518>
bitcoind = <pyln.testing.utils.BitcoinD object at 0x7f5bea5e4eb8>

    @unittest.skipIf(not EXPERIMENTAL_FEATURES, "upgrade protocol not available")
    @pytest.mark.developer("dev-force-features, dev-disconnect required")
    def test_upgrade_statickey_fail(node_factory, executor, bitcoind):
        """We reconnect at all points during retransmit, and we won't upgrade."""
        l1_disconnects = ['-WIRE_COMMITMENT_SIGNED',
                          '-WIRE_REVOKE_AND_ACK']
        l2_disconnects = ['-WIRE_REVOKE_AND_ACK',
                          '-WIRE_COMMITMENT_SIGNED']
    
        l1, l2 = node_factory.line_graph(2, opts=[{'may_reconnect': True,
                                                   'dev-no-reconnect': None,
                                                   'disconnect': l1_disconnects,
                                                   'dev-force-features': ["-13", "-21"],
                                                   # Don't have feerate changes!
                                                   'feerates': (7500, 7500, 7500, 7500)},
                                                  {'may_reconnect': True,
                                                   'dev-no-reconnect': None,
                                                   'disconnect': l2_disconnects,
                                                   'dev-disable-commit-after': 1}])
    
        # This HTLC will fail
        l1.rpc.sendpay([{'msatoshi': 1000, 'id': l2.info['id'], 'delay': 5, 'channel': '1x1x1'}], '00' * 32, payment_secret='00' * 32)
    
        # Each one should cause one disconnection, no upgrade.
        for d in l1_disconnects + l2_disconnects:
            l1.daemon.wait_for_log('Peer connection lost')
            l2.daemon.wait_for_log('Peer connection lost')
            assert not l1.daemon.is_in_log('option_static_remotekey enabled')
            assert not l2.daemon.is_in_log('option_static_remotekey enabled')
            l1.rpc.connect(l2.info['id'], 'localhost', l2.port)
    
        # On the last reconnect, it retransmitted revoke_and_ack.
        l1.daemon.wait_for_log('No upgrade: we retransmitted')
        l2.daemon.wait_for_log('No upgrade: pending changes')
    
        # Now when we reconnect, despite having an HTLC, we're quiescent.
        l1.rpc.disconnect(l2.info['id'], force=True)
        l1.rpc.connect(l2.info['id'], 'localhost', l2.port)
    
>       l1.daemon.wait_for_log('option_static_remotekey enabled at 2/2')

tests/test_connection.py:3653: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
contrib/pyln-testing/pyln/testing/utils.py:310: in wait_for_log
    return self.wait_for_logs([regex], timeout)
```

**Comment by rustyrussell** [comment]
Ack 886bede

**Pull request merged**